File: finetune-skill-otel/scripts/trace_grader.py
# /// script
# requires-python = ">=3.10"
# ///
"""
trace_grader.py — Stage 4: programmatic Jaccard grader for tool calls.

Deterministic scorer: (predicted_tool_call, ground_truth_tool_call) → [0.02, 1.0].
No LLM. Used by Stage 6 (probe), Stage 7 (cloud training), and Stage 8 (eval).

Formula (ToolRL-grounded, arXiv:2504.13958):

    r_name  = 1 if normalize_name(pred) == normalize_name(gt) else 0
    r_param = |keys(gt) ∩ keys(pred)| / |keys(gt) ∪ keys(pred)|   (Jaccard)
    r_value = |{ k ∈ intersect : gt[k] == pred[k] }|
    S_raw   = r_name + r_param + r_value
    S_max   = 1 + len(gt_args)                                    # name + one per arg
    score   = clamp(S_raw / S_max, floor=0.02, ceiling=1.0)

Wrong tool name → 0.02 floor (never 0.0 — zero-variance kills GRPO per the
`feedback_grader_no_zero_hard_gate` memory rule).

Parallel calls (Pattern D): per-call scores averaged × coverage penalty.

Refusal (empty call sets on both sides): 1.0 (correct refusal).

See `finetune-skill-otel/reference/trace-grader-reference.md` for the full
formula derivation, edge-case rules, and unit-test spec.
"""

from __future__ import annotations

from typing import Any


# ─── Normalization helpers ──────────────────────────────────────────────────


def normalize_name(name: str | None) -> str:
    """Case-insensitive, whitespace-trimmed name normalization."""
    if name is None:
        return ""
    return name.strip().lower()


def normalize_value(v: Any) -> Any:
    """
    Normalize a value for exact-match comparison.

    Rules (from trace-grader-reference.md):
      - None → None
      - bool → bool (not converted to int)
      - int/float → float (type-lenient: int(1) == float(1.0))
      - str → trimmed (case preserved)
      - list → tuple of normalized elements (order-preserving, hashable)
      - dict → tuple of sorted (key, value) pairs (order-independent)
      - other → passed through
    """
    if v is None:
        return None
    if isinstance(v, bool):
        # Tag booleans so they don't compare equal to numeric 0/1.
        # (Python's True == 1 is True by default, which would silently
        # accept `verified=1` as matching `verified=true` — wrong for
        # tool-routing where types matter.)
        return ("__bool__", v)
    if isinstance(v, (int, float)):
        return float(v)
    if isinstance(v, str):
        return v.strip()
    if isinstance(v, list):
        return tuple(normalize_value(x) for x in v)
    if isinstance(v, dict):
        return tuple(
            sorted(
                (normalize_value(k), normalize_value(vv)) for k, vv in v.items()
            )
        )
    return v


# ─── Jaccard + value matching primitives ────────────────────────────────────


def _jaccard_keys(a: dict, b: dict) -> float:
    """Jaccard overlap on the key sets. Empty-vs-empty = 1.0 (convention)."""
    keys_a = set(a.keys())
    keys_b = set(b.keys())
    union = keys_a | keys_b
    if not union:
        return 1.0  # both empty → trivially equal
    return len(keys_a & keys_b) / len(union)


def _count_value_matches(pred_args: dict, gt_args: dict) -> int:
    """Count keys in the intersection where normalized values are equal."""
    intersect = set(pred_args.keys()) & set(gt_args.keys())
    return sum(
        1
        for k in intersect
        if normalize_value(pred_args[k]) == normalize_value(gt_args[k])
    )


def _clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))


# ─── Single-call grader ─────────────────────────────────────────────────────


WRONG_TOOL_FLOOR = 0.02


def grade(pred: dict, gt: dict) -> float:
    """
    Score a single predicted tool call against ground truth.

    Args:
        pred: {"name": str, "arguments": dict}
        gt:   {"name": str, "arguments": dict}

    Returns:
        Score in [0.02, 1.0]. 0.02 floor for wrong tool name. 1.0 for
        perfect match.
    """
    pred_name = normalize_name(pred.get("name"))
    gt_name = normalize_name(gt.get("name"))

    # Wrong tool → floor. Never 0.0.
    if pred_name == "" or pred_name != gt_name:
        return WRONG_TOOL_FLOOR

    pred_args = pred.get("arguments") or {}
    gt_args = gt.get("arguments") or {}

    r_name = 1.0
    r_param = _jaccard_keys(pred_args, gt_args)
    r_value = _count_value_matches(pred_args, gt_args)

    s_raw = r_name + r_param + r_value
    s_max = 1 + len(gt_args)  # 1 for name + 1 per GT arg

    score = s_raw / s_max
    return _clamp(score, WRONG_TOOL_FLOOR, 1.0)


# ─── Parallel-call grader (Pattern D) ───────────────────────────────────────


def _best_match_idx(pred_call: dict, gt_calls: list[dict], taken: set[int]) -> int | None:
    """Find the index of the best-matching GT call that hasn't been consumed."""
    best_idx = None
    best_score = -1.0
    for i, gt_call in enumerate(gt_calls):
        if i in taken:
            continue
        s = grade(pred_call, gt_call)
        if s > best_score:
            best_score = s
            best_idx = i
    return best_idx


def grade_parallel(pred: dict, gt: dict) -> float:
    """
    Score a parallel tool-call set (Pattern D) against the ground truth.

    Uses per-call mean × coverage penalty: each predicted call is matched
    to its best-matching remaining GT call (greedy, GT consumed once per
    match), per-call scores are averaged, then multiplied by
    min(|pred|, |gt|) / max(|pred|, |gt|) to penalize missing or extra
    calls symmetrically.

    Refusal: empty-vs-empty call sets → 1.0 (correctly refused).
    Hallucinated call vs empty GT → 0.02 floor.
    Missed call vs non-empty GT → 0.02 floor.

    Args:
        pred: {"calls": [{"name": str, "arguments": dict}, ...]}
        gt:   {"calls": [{"name": str, "arguments": dict}, ...]}

    Returns:
        Score in [0.02, 1.0].
    """
    pred_calls = list(pred.get("calls") or [])
    gt_calls = list(gt.get("calls") or [])

    # Refusal cases
    if not pred_calls and not gt_calls:
        return 1.0
    if not pred_calls or not gt_calls:
        return WRONG_TOOL_FLOOR

    # Greedy best-match, consuming each GT call at most once
    taken: set[int] = set()
    per_call_scores: list[float] = []
    for p in pred_calls:
        idx = _best_match_idx(p, gt_calls, taken)
        if idx is None:
            per_call_scores.append(WRONG_TOOL_FLOOR)
        else:
            taken.add(idx)
            per_call_scores.append(grade(p, gt_calls[idx]))

    mean_per_call = sum(per_call_scores) / len(per_call_scores)
    coverage = min(len(pred_calls), len(gt_calls)) / max(
        len(pred_calls), len(gt_calls)
    )
    score = mean_per_call * coverage
    return _clamp(score, WRONG_TOOL_FLOOR, 1.0)


# ─── Mandatory grader sanity check ──────────────────────────────────────────


def _extract_tool_call_from_record(record: dict) -> dict | None:
    """Extract the GT tool call from a training record.

    Checks `ground_truth` field first (GRPO format: GT stored separately,
    messages end with user turn), then falls back to last assistant message
    (legacy), then `output` dict (test records).
    """
    import json as _json

    # GRPO format: ground_truth is a separate JSON field
    gt_raw = record.get("ground_truth")
    if gt_raw is not None:
        parsed = _json.loads(gt_raw) if isinstance(gt_raw, str) else gt_raw
        tcs = parsed if isinstance(parsed, list) else []
        if tcs:
            fn = tcs[0].get("function", {})
            args = fn.get("arguments")
            if isinstance(args, str):
                try:
                    args = _json.loads(args)
                except Exception:
                    args = {}
            return {"name": fn.get("name", ""), "arguments": args or {}}

    # Legacy: last assistant message in messages array
    for msg in reversed(record.get("messages") or []):
        if msg.get("role") != "assistant":
            continue
        tcs = msg.get("tool_calls") or []
        if tcs:
            fn = tcs[0].get("function", {})
            args = fn.get("arguments")
            if isinstance(args, str):
                try:
                    args = _json.loads(args)
                except Exception:
                    args = {}
            return {"name": fn.get("name", ""), "arguments": args or {}}

    # Legacy test format: output dict
    if "output" in record:
        return record["output"]
    return None


def sanity_check(gt_record: dict) -> None:
    """
    Run the mandatory grader sanity check against a real training record.

    Must be called once per workflow before training starts. Verifies:
      1. Self-match scores 1.0
      2. Wrong tool scores the 0.02 floor
      3. Partial args (half of GT keys) score in [0.2, 0.8]

    Accepts both formats:
      - New: {"messages": [...], "tools": [...]} (from otel_distill.py)
      - Legacy: {"output": {"name": ..., "arguments": {...}}}

    Raises AssertionError if any property fails. A failing sanity check
    indicates a grader bug (case sensitivity, normalization error) that
    would silently degrade training.
    """
    output = _extract_tool_call_from_record(gt_record)
    if output is None:
        raise ValueError(
            "sanity_check: cannot extract tool call from record — "
            "expected 'messages' with assistant tool_calls or legacy 'output' key"
        )

    # 1. Self-match
    self_score = grade(output, output)
    assert self_score == 1.0, f"Self-match score must be 1.0, got {self_score}"

    # 2. Wrong tool floor
    wrong = {**output, "name": "__intentionally_wrong_tool_name__"}
    wrong_score = grade(wrong, output)
    assert wrong_score == WRONG_TOOL_FLOOR, (
        f"Wrong-tool score must be {WRONG_TOOL_FLOOR}, got {wrong_score}"
    )

    # 3. Partial args in (0.2, 0.8)
    args = output.get("arguments") or {}
    if len(args) >= 2:
        partial_args = {
            k: v for i, (k, v) in enumerate(args.items()) if i < len(args) // 2
        }
        partial = {**output, "arguments": partial_args}
        partial_score = grade(partial, output)
        assert 0.2 <= partial_score <= 0.9, (
            f"Partial-args score out of (0.2, 0.9) range: {partial_score}"
        )

File: finetune-skill-otel/scripts/trace_grader.js
// trace_grader.js — Stage 4: programmatic Jaccard grader for tool calls.
//
// Deterministic scorer: (predicted_tool_call, ground_truth_tool_call) → [floor, 1.0].
// No LLM. Used by cloud training (Stage 7) and cloud eval (Stage 8).
//
// Formula (ToolRL-grounded, arXiv:2504.13958):
//
//     r_name  = 1 if normalize_name(pred) == normalize_name(gt) else 0
//     r_param = |keys(gt) ∩ keys(pred)| / |keys(gt) ∪ keys(pred)|   (Jaccard)
//     r_value = |{ k ∈ intersect : gt[k] == pred[k] }|
//     S_raw   = r_name + r_param + r_value
//     S_max   = 1 + len(gt_args)                                    # name + one per arg
//     score   = clamp(S_raw / S_max, floor=0.02, ceiling=1.0)
//
// Wrong tool name → 0.02 floor (never 0.0 — zero-variance kills GRPO
// per feedback_grader_no_zero_hard_gate: a grader returning 0.0 for
// attempted answers causes 80% zero-variance prompts, producing flat
// training. Use 0.02 minimum so the model always gets a nonzero gradient
// signal — arXiv:2509.21880 "No Prompt Left Behind" confirms zero-variance
// prompt frequency is already 30–99% per batch without floor issues).
//
// GRADER_CONFIG is injected as a global by the publish step or cloud runner.
// Shape: { wrong_tool_floor, case_insensitive_params, tool_schema, formula_version }

// When embedded by the publish step, GRADER_CONFIG is declared above this file.
// When running standalone (e.g. tests), it may not exist — fall back to defaults.
// Using var (not const) so the publish step's `const GRADER_CONFIG = ...` above
// doesn't cause a duplicate-declaration error when this file is concatenated.
// eslint-disable-next-line no-undef, no-var
var CONFIG = typeof GRADER_CONFIG !== "undefined" ? GRADER_CONFIG : {};
var WRONG_TOOL_FLOOR = CONFIG.wrong_tool_floor != null ? CONFIG.wrong_tool_floor : 0.02;

// Runtime assertion: warn if GRADER_CONFIG was not injected (e.g. cloud
// sandbox didn't prepend it). The grader still works (floor=0.02) but
// case_insensitive_params will be empty, losing enum-sensitivity.
if (!CONFIG.tool_schema || CONFIG.tool_schema.length === 0) {
  if (typeof console !== "undefined" && console.warn) {
    console.warn(
      "[trace_grader.js] GRADER_CONFIG.tool_schema is empty — " +
      "case_insensitive_params will not apply. Was GRADER_CONFIG prepended?"
    );
  }
}

// Build a Set of "tool|param" keys for O(1) lookup of case-insensitive params.
// These are string-enum parameters (e.g. HTTP method, category) where
// "GET" should match "get" — detected at build time by trace_grader_builder.py.
// eslint-disable-next-line no-var
var CASE_INSENSITIVE_SET = new Set(
  (CONFIG.case_insensitive_params || []).map(
    (entry) => `${(entry.tool || "").toLowerCase()}|${entry.param}`
  )
);


// ─── Normalization helpers ──────────────────────────────────────────────────

/**
 * Case-insensitive, whitespace-trimmed name normalization.
 * Matches Python: name.strip().lower()
 */
function normalizeName(name) {
  if (name == null) return "";
  return String(name).trim().toLowerCase();
}

/**
 * Check if a param is in the case-insensitive set for a given tool.
 */
function isCaseInsensitiveParam(toolName, paramName) {
  const key = `${normalizeName(toolName)}|${paramName}`;
  return CASE_INSENSITIVE_SET.has(key);
}

/**
 * Normalize a value for exact-match comparison.
 *
 * Rules (matching trace_grader.py exactly):
 *   - null/undefined → null
 *   - boolean → tagged tuple ["__bool__", value] (so true !== 1)
 *   - number → number (int/float are the same type in JS, no issue)
 *   - string → trimmed (case preserved by default)
 *   - array → JSON of normalized elements (order-preserving)
 *   - object → JSON of sorted [key, value] pairs (order-independent)
 *   - other → passed through
 *
 * Python gotcha: Python's `True == 1` is `True`. JavaScript's `true === 1`
 * is `false`, but `true == 1` is `true`. We tag booleans to match the Python
 * behavior of distinguishing bool from int, which matters for tool routing
 * (e.g. `verified=true` should not match `verified=1`).
 *
 * Returns a string representation for reliable equality comparison, since
 * JS objects/arrays can't be compared with ===.
 */
function normalizeValue(v) {
  if (v === null || v === undefined) {
    return "__null__";
  }
  if (typeof v === "boolean") {
    // Tag booleans so they don't compare equal to numeric 0/1.
    // Matches Python's ("__bool__", v) tagging.
    return `__bool__:${v}`;
  }
  if (typeof v === "number") {
    return `__num__:${v}`;
  }
  if (typeof v === "string") {
    return `__str__:${v.trim()}`;
  }
  if (Array.isArray(v)) {
    // Order-preserving — [1, 2] !== [2, 1]
    const items = v.map((x) => normalizeValue(x));
    return `__arr__:[${items.join(",")}]`;
  }
  if (typeof v === "object") {
    // Order-independent — sort by key
    const entries = Object.keys(v)
      .sort()
      .map((k) => `${normalizeValue(k)}=${normalizeValue(v[k])}`);
    return `__obj__:{${entries.join(",")}}`;
  }
  return `__other__:${String(v)}`;
}

/**
 * Compare two values for equality after normalization.
 * Optionally applies case-insensitive string comparison for enum params.
 */
function valuesEqual(predVal, gtVal, caseInsensitive) {
  if (caseInsensitive && typeof predVal === "string" && typeof gtVal === "string") {
    return normalizeValue(predVal.toLowerCase()) === normalizeValue(gtVal.toLowerCase());
  }
  return normalizeValue(predVal) === normalizeValue(gtVal);
}


// ─── Jaccard + value matching primitives ────────────────────────────────────

/**
 * Jaccard overlap on the key sets. Empty-vs-empty = 1.0 (convention).
 */
function jaccardKeys(a, b) {
  const keysA = new Set(Object.keys(a));
  const keysB = new Set(Object.keys(b));

  const unionSize = new Set([...keysA, ...keysB]).size;
  if (unionSize === 0) return 1.0; // both empty → trivially equal

  let intersectCount = 0;
  for (const k of keysA) {
    if (keysB.has(k)) intersectCount++;
  }
  return intersectCount / unionSize;
}

/**
 * Count keys in the intersection where normalized values are equal.
 */
function countValueMatches(predArgs, gtArgs, toolName) {
  const predKeys = new Set(Object.keys(predArgs));
  let count = 0;
  for (const k of Object.keys(gtArgs)) {
    if (!predKeys.has(k)) continue;
    const ci = isCaseInsensitiveParam(toolName, k);
    if (valuesEqual(predArgs[k], gtArgs[k], ci)) {
      count++;
    }
  }
  return count;
}

function clamp(v, lo, hi) {
  return Math.max(lo, Math.min(hi, v));
}


// ─── Single-call grader ─────────────────────────────────────────────────────

/**
 * Score a single predicted tool call against ground truth.
 *
 * @param {object} pred - {name: string, arguments: object}
 * @param {object} gt   - {name: string, arguments: object}
 * @returns {number} Score in [WRONG_TOOL_FLOOR, 1.0]
 */
function gradeSingle(pred, gt) {
  const predName = normalizeName(pred.name);
  const gtName = normalizeName(gt.name);

  // Wrong tool → floor. Never 0.0.
  if (predName === "" || predName !== gtName) {
    return WRONG_TOOL_FLOOR;
  }

  const predArgs = pred.arguments || {};
  const gtArgs = gt.arguments || {};

  const rName = 1.0;
  const rParam = jaccardKeys(predArgs, gtArgs);
  const rValue = countValueMatches(predArgs, gtArgs, gt.name);

  const sRaw = rName + rParam + rValue;
  const sMax = 1 + Object.keys(gtArgs).length; // 1 for name + 1 per GT arg

  const score = sRaw / sMax;
  return clamp(score, WRONG_TOOL_FLOOR, 1.0);
}


// ─── Parallel-call grader (Pattern D) ───────────────────────────────────────

/**
 * Find the index of the best-matching GT call that hasn't been consumed.
 */
function bestMatchIdx(predCall, gtCalls, taken) {
  let bestIdx = null;
  let bestScore = -1.0;
  for (let i = 0; i < gtCalls.length; i++) {
    if (taken.has(i)) continue;
    const s = gradeSingle(predCall, gtCalls[i]);
    if (s > bestScore) {
      bestScore = s;
      bestIdx = i;
    }
  }
  return bestIdx;
}

/**
 * Score a parallel tool-call set (Pattern D) against the ground truth.
 *
 * Uses per-call mean × coverage penalty: each predicted call is matched
 * to its best-matching remaining GT call (greedy, GT consumed once per
 * match), per-call scores are averaged, then multiplied by
 * min(|pred|, |gt|) / max(|pred|, |gt|) to penalize missing or extra
 * calls symmetrically.
 *
 * Refusal: empty-vs-empty call sets → 1.0 (correctly refused).
 * Hallucinated call vs empty GT → floor.
 * Missed call vs non-empty GT → floor.
 */
function gradeParallel(predCalls, gtCalls) {
  // Refusal cases
  if (predCalls.length === 0 && gtCalls.length === 0) return 1.0;
  if (predCalls.length === 0 || gtCalls.length === 0) return WRONG_TOOL_FLOOR;

  // Greedy best-match, consuming each GT call at most once
  const taken = new Set();
  const perCallScores = [];
  for (const p of predCalls) {
    const idx = bestMatchIdx(p, gtCalls, taken);
    if (idx === null) {
      perCallScores.push(WRONG_TOOL_FLOOR);
    } else {
      taken.add(idx);
      perCallScores.push(gradeSingle(p, gtCalls[idx]));
    }
  }

  const meanPerCall =
    perCallScores.reduce((sum, s) => sum + s, 0) / perCallScores.length;
  const coverage =
    Math.min(predCalls.length, gtCalls.length) /
    Math.max(predCalls.length, gtCalls.length);
  const score = meanPerCall * coverage;
  return clamp(score, WRONG_TOOL_FLOOR, 1.0);
}


// ─── Message parser ─────────────────────────────────────────────────────────

/**
 * Extract tool calls from an OpenAI chat-completion assistant message.
 *
 * Input shape (from cloud evaluator):
 *   {
 *     role: "assistant",
 *     tool_calls: [{
 *       type: "function",
 *       function: { name: "search", arguments: '{"q": "tablet"}' }
 *     }]
 *   }
 *
 * Returns array of {name, arguments} objects.
 * Handles: missing tool_calls, non-array tool_calls, unparseable arguments.
 */
function extractToolCalls(message) {
  if (!message || typeof message !== "object") return [];

  const rawCalls = message.tool_calls;
  if (!Array.isArray(rawCalls)) return [];

  const result = [];
  for (const call of rawCalls) {
    if (!call || typeof call !== "object") continue;

    const fn = call.function || call;
    const name = fn.name || null;

    let args = fn.arguments;
    if (typeof args === "string") {
      try {
        args = JSON.parse(args);
      } catch {
        // Unparseable arguments → treat as empty (wrong values will reduce score)
        args = {};
      }
    }
    if (args == null || typeof args !== "object" || Array.isArray(args)) {
      args = {};
    }

    result.push({ name, arguments: args });
  }
  return result;
}


// ─── Main grade function (cloud evaluator entry point) ──────────────────────

/**
 * Grade a predicted assistant message against the expected ground truth.
 *
 * This is the entry point called by the LangDB Cloud evaluator.
 * Both `predicted` and `expected` are OpenAI chat-completion assistant
 * messages with tool_calls arrays.
 *
 * @param {object} predicted - The model's output message
 * @param {object} expected  - The ground truth message
 * @returns {{ score: number, reason: string }}
 */
function grade(predicted, expected) {
  // Extract tool calls from both messages
  const predCalls = extractToolCalls(predicted);
  const gtCalls = extractToolCalls(expected);

  // Single-call fast path (most common case)
  if (predCalls.length <= 1 && gtCalls.length <= 1) {
    if (predCalls.length === 0 && gtCalls.length === 0) {
      return { score: 1.0, reason: "correct refusal: no tool calls expected or predicted" };
    }
    if (predCalls.length === 0) {
      return {
        score: WRONG_TOOL_FLOOR,
        reason: "missed call: expected tool call but model produced none",
      };
    }
    if (gtCalls.length === 0) {
      return {
        score: WRONG_TOOL_FLOOR,
        reason: `hallucinated call: model called ${predCalls[0].name} but no call expected`,
      };
    }

    const score = gradeSingle(predCalls[0], gtCalls[0]);
    const pred = predCalls[0];
    const gt = gtCalls[0];

    if (score >= 1.0) {
      return { score: 1.0, reason: `perfect match: ${gt.name}` };
    }
    if (normalizeName(pred.name) !== normalizeName(gt.name)) {
      return {
        score,
        reason: `wrong tool: predicted ${pred.name}, expected ${gt.name}`,
      };
    }
    return {
      score: Math.round(score * 1000) / 1000,
      reason: `partial match on ${gt.name}: score=${score.toFixed(3)}`,
    };
  }

  // Multi-call path (Pattern D: parallel tool calls)
  const score = gradeParallel(predCalls, gtCalls);

  if (score >= 1.0) {
    return {
      score: 1.0,
      reason: `perfect parallel match: ${gtCalls.length} calls`,
    };
  }
  return {
    score: Math.round(score * 1000) / 1000,
    reason: `parallel match: ${predCalls.length} predicted vs ${gtCalls.length} expected, score=${score.toFixed(3)}`,
  };
}
